fs-cache/src/cache.rs

```diff
@@ -1,60 +1,489 @@
             label,
             max_memory_bytes
         );
-        Ok(Self {
-            storage: MemoryLimitedStorage::new(label, path, max_memory_bytes)?,
-        })
+
+        cache.load_fs()?;
+        Ok(cache)
     }
 
+    /// Retrieves a value by key from memory cache or disk, returns None if not found.
     pub fn get(&mut self, key: &K) -> Option<V> {
-        let result = self.storage.get(key);
-        log::debug!(
-            "{} cache: get key={} -> found={}",
-            self.storage.label(),
-            key,
-            result.is_some()
-        );
-        result
+        log::debug!("cache/{}: retrieving value for key {}", self.label, key);
+
+        if let Some(value) = self.get_from_memory(key) {
+            return Some(value);
+        }
+
+        self.get_from_disk(key)
     }
 
+    /// Stores a value with the given key, writing to disk and updating memory cache.
     pub fn set(&mut self, key: K, value: V) -> Result<()> {
+        log::debug!("cache/{}: setting value for key {}", self.label, key);
         // Check if value already exists
-        if self.storage.get(&key).is_some() {
-            log::debug!(
-                "{} cache: skip setting existing key={}",
-                self.storage.label(),
-                key
-            );
+        if self.get(&key).is_some() {
+            log::debug!("cache/{}: skipping existing key {}", self.label, key);
             return Ok(());
         }
 
-        log::debug!("{} cache: set key={}", self.storage.label(), key);
-        self.storage.set(key, value)
+        // Always write to disk first
+        self.write_to_disk(&key, &value)?;
+
+        // Then update memory cache
+        self.cache_in_memory(&key, &value)?;
+
+        log::debug!("cache/{}: set key={}", self.label, key);
+        Ok(())
+    }
+
+    fn load_fs(&mut self) -> Result<()> {
+        if !self.path.exists() {
+            return Err(ArklibError::Storage(
+                self.label.clone(),
+                "Folder does not exist".to_owned(),
+            ));
+        }
+
+        if !self.path.is_dir() {
+            return Err(ArklibError::Storage(
+                self.label.clone(),
+                "Path is not a directory".to_owned(),
+            ));
+        }
+
+        // First pass: collect metadata only
+        let mut file_metadata = Vec::new();
+        for entry in fs::read_dir(&self.path)? {
+            let path = entry?.path();
+            if path.is_file() {
+                let key: K =
+                    extract_key_from_file_path(&self.label, &path, true)?;
+                file_metadata.push((key.clone(), self.get_file_size(&key)?));
+            }
+        }
+
+        // Sort by size before loading
+        file_metadata.sort_by(|a, b| b.1.cmp(&a.1));
+
+        // Clear existing cache
+        self.memory_cache.clear();
+        self.current_memory_bytes = 0;
+
+        // Load files that fit in memory
```

Comment:
Interesting approach with pre-loading the cache. We probably should have a switch, so the app developer could turn it on/off.

Reply:
Done!